### PR TITLE
🏃 Add and fix shellcheck errors

### LINKED ISSUE #1551: Add shellcheck to make verify
As part of `make verify` target, we should also run shellcheck on all the scripts in this repository. An example is available in the CABPK repository https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/blob/master/hack/verify-shellcheck.sh

/kind feature
/help
/milestone v0.3.0
/priority important-longterm

### PR DESCRIPTION
**Pull request by ashish-amarnath**
Signed-off-by: Ashish Amarnath <ashish.amarnath@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1551

### CODE REVIEW

File: hack/verify-shellcheck.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+wget https://storage.googleapis.com/shellcheck/"${DOWNLOAD_FILE}"
+tar xf "${DOWNLOAD_FILE}"
+cd "${VERSION}" || exit
+
+echo "Running shellcheck..."
+cd "${ROOT_PATH}" || exit
+OUT="${TMP_DIR}/out.log"
+IGNORE_FILES=$(find . -name "*.sh" | grep third_party)
+echo "Ignoring shellcheck on ${IGNORE_FILES}"
+FILES=$(find . -name "*.sh" | grep -v third_party)
```
Comment (ashish-amarnath) [inline hack/verify-shellcheck.sh]: This file is mostly lifted from the previous cabpk repo except for lines 45, 46 and 47

File: scripts/ci-integration.sh [hunk 2]
```diff
@@ -19,7 +19,7 @@ set -o nounset
 set -o pipefail
 
 REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
-# shellcheck source=../hack/ensure-go.sh
+# shellcheck source=/dev/null
```
Comment (detiber) [inline scripts/ci-integration.sh]: I would expect this to point to an actual file not /dev/null

Comment (detiber) [reply scripts/ci-integration.sh]: Same comment applies to other places where `shellcheck source=/dev/null`

File: hack/utils.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
```
Comment (ashish-amarnath) [inline hack/utils.sh]: This file was lifted from cabpk repo

File: scripts/ci-integration.sh (hunk 2 above)
Comment (ashish-amarnath) [inline scripts/ci-integration.sh]: This causes SC1091 https://github.com/koalaman/shellcheck/wiki/SC1091
IMO, it is OK to ignore this as we run `shellcheck` on all `.sh` files in the repo.
Also, the correct way to ignore this seems to be like
```
# shellcheck disable=SC1091
```
However, the `/dev/null` approach seems like the prefered way in the repo.

LMK if the way I've understood this is not correct.

Comment (detiber) [reply scripts/ci-integration.sh]: Reading that, I'm wondering if we should be invoking `shellcheck` with the `-x` flag to allow it to follow the file.

Comment (ashish-amarnath) [reply scripts/ci-integration.sh]: I tried that, but it didn't work.

Comment (ashish-amarnath) [reply scripts/ci-integration.sh]: I was able to get that to work. The last time I tried the `-x` I wasn't providing the path to the sourced file correctly.

Comment (ashish-amarnath) [reply scripts/ci-integration.sh]: I was doing this wrong. I've fixed it now.

Comment (ashish-amarnath) [reply scripts/ci-integration.sh]: I've fixed it now.

File: hack/verify-shellcheck.sh (hunk 1 above)
Comment (ashish-amarnath) [reply hack/verify-shellcheck.sh]: :running:

File: hack/verify-shellcheck.sh [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+# cleanup on exit
+cleanup() {
+  echo "Cleaning up..."
+  rm -rf "${TMP_DIR}"
+}
+trap cleanup EXIT
+
+# install shellcheck (Linux-x64 only!)
+cd "${TMP_DIR}" || exit
+VERSION="shellcheck-v0.6.0"
```
Comment (vincepri) [inline hack/verify-shellcheck.sh]: 0.7.0 should be available, can we use that one instead?

Comment (ashish-amarnath) [reply hack/verify-shellcheck.sh]: Yes, we sure can!

**Comment by ashish-amarnath** [comment]
This check should not run on `darwin`

**Comment by vincepri** [comment]
Instead of excluding it on Mac, can you just not install shellcheck if it's present on path? And run it as usual otherwise?

**Comment by ashish-amarnath** [comment]
I didn't exclude it on Mac. `shellcheck` for mac also works. So I made it download the binary for Mac. 
Not sure why this was excluded earlier.

**Comment by vincepri** [comment]
Let's avoid messing with folks' paths if it's not needed, if shellcheck is already available in path, we should skip the download imo

**Comment by detiber** [comment]
> Let's avoid messing with folks' paths if it's not needed, if shellcheck is already available in path, we should skip the download imo

If that's the case, should we ensure that the existing `shellcheck` is >= the expected version?

**Comment by ashish-amarnath** [comment]
`shellcheck` gets downloaded into a tmp dir that gets cleaned up, so we will not be messing w/ `PATH`. So we can give folks a consistent behavior locally and in ci.

**Comment by ashish-amarnath** [comment]
https://github.com/kubernetes-sigs/cluster-api/pull/1621/files#diff-21849b598a2db215f292c1bd4620e454R32

File: hack/verify-shellcheck.sh [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  echo "Cleaning up..."
+  rm -rf "${TMP_DIR}"
+}
+trap cleanup EXIT
+
+# install shellcheck (x64 only!)
+cd "${TMP_DIR}" || exit
+VERSION="shellcheck-v0.7.0"
+DOWNLOAD_FILE="${VERSION}.${OS}.x86_64.tar.xz"
+wget https://storage.googleapis.com/shellcheck/"${DOWNLOAD_FILE}"
```
Comment (chuckha) [inline hack/verify-shellcheck.sh]: hah, my system didn't have wget. Isn't bash amazing.

Can you add a failure case if wget doesn't exist on the system with exit 1 saying `wget` is missing?

**Comment by chuckha** [comment]
lgtm pending one wget issue. wget is not on os x by default.

File: hack/verify-shellcheck.sh (hunk 5 above)
Comment (ncdc) [reply hack/verify-shellcheck.sh]: Let's use curl, please 😄

Comment (chuckha) [reply hack/verify-shellcheck.sh]: that is also acceptable :D

Comment (ashish-amarnath) [reply hack/verify-shellcheck.sh]: done. ptal

**Comment by chuckha** [comment]
that worked swimmingly. I'll lgtm when the tests pass.

**Pull request merged**